[ fix #4572 ] added piSort UnivSort funSort and Setω+n to the user-manual

pnlph opened the pull request:
Finally, I was editing the page of Universe levels. I think its name should also be changed maybe to `Universe hierarchy`, `Sort hierarchy` or something similar? What do you think? Since the arrival of `Setω` the levels are not the only possible universes.

pnlph committed:
added piSort UnivSort funSort to user-manual

andreasabel commented on doc/user-manual/language/universe-levels.lagda.rst:
```diff
@@ -260,6 +260,58 @@ Level) → Set n`` belongs to ``Setω`` which is not of this form. The
+where:
+
+* ``_5`` is a metavariable that represents the sort of ``A``.
+* ``funSort _5 _5)`` is the sort of ``A → A``
```
Extra ).

andreasabel commented on doc/user-manual/language/universe-levels.lagda.rst:
```diff
@@ -260,6 +260,58 @@ Level) → Set n`` belongs to ``Setω`` which is not of this form. The
+
+More examples:
+
+* ``piSort Level (\l. Set l)`` evaluates to ``Setω``
```
Please use `λ` and `→` in the notation for lambda-abstraction.

andreasabel reviewed (changes requested):
Found some typos.

jespercockx commented:
I'd suggest this information should go on its own page about Agda's sort system, since it does not have anything to do with universe levels *per se*.

jespercockx commented:
I think it would also make sense to write down the behaviour of `funSort`, `piSort`, and `univSort` on closed sorts explicitly, i.e.
```
funSort Setω     _        = Setω
funSort _        Setω     = Setω
funSort (Set a)  (Set b)  = Set (a ⊔ b)
funSort (Prop a) (Set b)  = Set (a ⊔ b)
funSort (Set a)  (Prop b) = Prop (a ⊔ b)
funSort (Prop a) (Prop b) = Prop (a ⊔ b)

piSort s1 (\x -> s2) = funSort s1 s2          (if x does not occur freely in s2)
piSort s1 (\x -> s2) = Setω                   (if x occurs freely in s2)

univSort (Set a)  = Set (lsuc a)
univSort (Prop a) = Set (lsuc a)
univSort Setω     = Setω                      (only if --omega-in-omega is enabled)
```

jespercockx commented on doc/user-manual/language/universe-levels.lagda.rst:
```diff
@@ -260,6 +260,58 @@ Level) → Set n`` belongs to ``Setω`` which is not of this form. The
+  ``piSort (univSort _9) (λ A → funSort (univSort _7)
+  (funSort _7 (funSort _9 _7)))``
+
+Note that ``funSort``, ``PiSort`` and ``UnivSort`` are not always
```
With the current sort system, `funSort` and `piSort` are total functions on sort, but `univSort` is not (unless `--omega-in-omega` is enabled). Any uses of `univSort` will lead to a 'has bigger sort' constraint that ensures the argument is not `Setω`.

pnlph replied:
OK, I changed the paragraph, please take a look.

pnlph committed:
Update universe-levels.lagda.rst

Typos found by Andreas

gallais commented on doc/user-manual/language/universe-levels.lagda.rst:
```diff
@@ -300,8 +300,8 @@ to the right sort once their arguments are known.
-* ``piSort Level (\l. Set l)`` evaluates to ``Setω``
-* ``piSort (Set l) (\_. Set l')`` evaluates to ``Set (l ⊔ l')``
+* ``piSort Level (λ → Set l)`` evaluates to ``Setω``
+* ``piSort (Set l) (λ → Set l')`` evaluates to ``Set (l ⊔ l')``
```
Looks like we are missing a bound `l` in the first case and an underscore in the second.

pnlph replied:
Thanks!

pnlph committed:
added sort-system to doc

pnlph commented:
I added a new `sort-system` page to the documentation, like Jesper proposed. I added it also to the index of the language folder, I hope this suffices...

There are code bits describing the behaviour as proposed, introduced by `.. code-block::` but without `agda`. I hope they are not being typechecked. Is there another way of introducing dummy code?

andreasabel commented on doc/user-manual/language/sort-system.lagda.rst:
```diff
@@ -0,0 +1,112 @@
+compute these sort metavariables, just like it does when computing
+regular term metavariables.
+
+But this sort metavariable need other constructors to solve function
```
need --> needs

The sentence is a bit unclear.

I think there should be a basic explanation that if in function type `A -> B` domain `A` has sort `sA` and codomain `B` has sort `sB`, then the function type has sort `funSort sA sB`. If `sA` and `sB` are known, `funSort sA sB` can be computed to a sort value.


Like there should be a basic explanation that if in dependent function type `(x : A) -> B` domain `A` has sort `sA` and codomain `B` has sort `sB` dependent on `x : A`, then the function type has sort `piSort sA (\ x -> sB)`.

pnlph replied:
> need --> needs

Thanks, this happens when committing at night...

> I think there should be a basic explanation...

I agree, please see how it reads now.

andreasabel reviewed (changes requested):
Thanks.  Looks good in general, I have one request, though.

pnlph committed:
additional explanations on funSort and PiSort

andreasabel commented on doc/user-manual/language/sort-system.lagda.rst:
```diff
@@ -0,0 +1,150 @@
+* ``sA`` and ``sB`` are two (possibly different) sorts.
+* ``A : sA``, meaning that ``A`` is a type that has sort ``sA``.
+* ``x : A``, meaning that ``x`` has type ``A``.
+* ``B : sB``, meaning that B is a type (possibly different than ``A``) that
```
B --> ``B``  sB --> ``sB``

andreasabel commented on doc/user-manual/language/sort-system.lagda.rst:
```diff
@@ -0,0 +1,150 @@
+  has sort sB.
+
+Under these conditions, we can build the dependent function type
+``(x : A) → B : PiSort sA (λ x → sB)``. This type signature means that the
```
PiSort --> piSort  (several times)

andreasabel commented on doc/user-manual/language/sort-system.lagda.rst:
```diff
@@ -0,0 +1,150 @@
+  ``piSort (univSort _9) (λ A → funSort (univSort _7)
+  (funSort _7 (funSort _9 _7)))``
+
+Note that ``funSort`` and ``PiSort`` are total functions on sort. But
```
PiSort --> piSort

andreasabel commented on doc/user-manual/language/sort-system.lagda.rst:
```diff
@@ -0,0 +1,150 @@
+  (funSort _7 (funSort _9 _7)))``
+
+Note that ``funSort`` and ``PiSort`` are total functions on sort. But
+``UnivSort`` is not always well-defined. Eg. without adding the
```
UnivSort --> univSort

andreasabel commented on doc/user-manual/language/sort-system.lagda.rst:
```diff
@@ -0,0 +1,150 @@
+regular term metavariables.
+
+But this sort metavariable needs other constructors to solve function
+types. The constructor ``funSort`` computes the sort of a function type
```
It should be made very clear that the sort constructors `funSort` `univSort` `piSort` are internal constructors that are sometimes printed, but can not be entered by the user.

pnlph replied:
I added a note before the explanations.

andreasabel reviewed (changes requested):
Getting very good.
- Some clarification still need.
- Some case issues.

pnlph committed:
piSort and comments - minor

pnlph committed:
case issue and note internal constructors

jespercockx commented:
I just merged #4595 so now the documentation should also mention the new sorts `Setωᵢ`. This also means the `Has bigger sort` constraint is now no longer necessary and has been removed. Do you want to update the documentation in this PR or should I do it directly?

pnlph commented:
I will see how the changelog looks like, and I will update here.

pnlph committed:
added Setω+n to sort system in doc

Documentation for PR #4609

pnlph committed:
added names for Setω+n to lexical structures in docs

Also updated links in sorts system page.

andreasabel commented on doc/user-manual/language/sort-system.lagda.rst:
```diff
@@ -0,0 +1,193 @@
+---------------
+
+Agda implements sorts of the form ``Setωᵢ``, where ``i`` is a natural number
+``i : Nat``. These sorts constitute a second hierarchy ``Setωᵢ : Setωᵢ₊₁``
```
Is really `i : Nat`?  This looks like `i` can be any Agda-expression of type `Nat` (as in `BUILTIN NATURAL Nat`).  Isn't it rather that `i` is a numeral in decimal representation, i.e. `0`, `1`, ... etc.?

andreasabel commented on doc/user-manual/language/sort-system.lagda.rst:
```diff
@@ -0,0 +1,193 @@
+expression ``∀ {i} (A : Setω i) → A → A`` would not be a well-formed agda
+term.
+
+You can also refer to these sorts with the alternative syntax ``Setωi``.
```
Is `Setω0` also legal (and the one where `0` is a subscript)?

pnlph replied:
I would say so, but let's ask @jespercockx to double-check: are `Setω0` and `Setω₀` legal keywords after #4609 was merged? They behave like their counterparts `Set0` and `Set₀` , right?

jespercockx replied:
They are legal but they are not keywords, you can rename `Setω` to anything such as `BleeBlorkl` in which case you can write `BleeBlorkl0` or `BleeBlorkl₀` and they mean the same thing as `BleeBlorkl` (which is to say, `Setω`).

pnlph replied:
OK, then if `Set`, `Setω` and all its derivations are not keywords, I have to remove them from the list of keywords in [lexical structure](https://agda.readthedocs.io/en/latest/language/lexical-structure.html), right?

jespercockx replied:
Ah yes, I forgot to update that list in my PR. `Set` and `Prop` should indeed be removed from that list, though you don't necessarily have to add it to this PR (as @andreasabel said, it is better to keep PR's focused on one thing).

pnlph replied:
OK perfect, i reverted the inclusion of `Setω` as keyword, you can do it for `Set` in yours.

andreasabel reviewed (changes requested):
Thanks for fixing my issues and for the new material on Set(omega)i.

As you can see, I have also issues with the new material. 😉 

For the sake of finally merging this PR, may I suggest to only fix issues and not add further new material?  New material (that is not strictly related) can go into a new PR.  Otherwise, I am afraid we will play this ping-pong game between contributor and reviewer for a long time... 😉 

UPDATE: I saw @jespercockx requested the new material.  Oh well, you are in a double bind between two bosses pulling into different directions. 🤣

andreasabel commented:
Note: the `user-manual-test` fails with this PR: https://travis-ci.org/github/agda/agda/jobs/690102891
```
The name of the top level module does not match the file name. 
The module sort.system should be defined in one of the following files: ...
```

pnlph committed:
updated `i` is a natural number expression

pnlph commented:
> This is not correct, the `i` can _only_ be a natural number literal, not an arbitrary expression of type `Nat` (how would that even work?)

OK, I thought that `Setω+n` would also work with arbitrary expressions of type `Nat` that may contain free variables _via_ `funSort`, `univSort` and `piSort`... or different constructors `funSortω`, `univSortω` and `piSortω` must be implemented for this to work?

jespercockx commented:
> 
> 
> > This is not correct, the `i` can _only_ be a natural number literal, not an arbitrary expression of type `Nat` (how would that even work?)
> 
> OK, I thought that `Setω+n` would also work with arbitrary expressions of type `Nat` that may contain free variables _via_ `funSort`, `univSort` and `piSort`... or different constructors `funSortω`, `univSortω` and `piSortω` must be implemented for this to work?

No, the index of `Setωi` cannot refer to any free variables, it is really just a literal natural number. I'm not sure why you would need `funSortω`, `univSortω` and `piSortω`, the normal `funSort`, `univSort`, and `piSort` still work for `Setωi`. For example, `piSort s (\x -> Setωi)` will always reduce to `funSort s Setωi` since there is no way for the `i` to depend on the variable `x`.

To put it in other words: the fact that `Setωi` cannot depend on any variables is an intentional design decision. Allowing that would necessitate an additional universe `Set2ω`, which would then soon lead to `Set2ω+1`, etc, which we'd like to avoid.

pnlph committed:
`i` can only be a literal natural number

Also reverted the addition of Setω as keyword in lexical-structure.

andreasabel reviewed (approved):
Looks good now.

andreasabel commented:
Thanks @pnlph !
